### PR TITLE
Optimize LogEventListeningService and LogEventProcessors.

### LINKED ISSUE #2632: Get -> Modify -> Set is slow, consider collect all logEvents then generate the t...
https://github.com/AElfProject/AElf/blob/d0f009f1d1ca9b20670677f75423012e573a1550/src/AElf.Kernel.SmartContract.ExecutionPluginForMethodFee/TransactionFeeChargedLogEventProcessor.cs#L52-L57

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a `TODO` comment in d0f009f1d1ca9b20670677f75423012e573a1550 when #2611 was merged. cc @AElfProject.

### LINKED ISSUE #2633: Get -> Modify -> Set is slow, consider collect all logEvents then generate the t...
https://github.com/AElfProject/AElf/blob/d0f009f1d1ca9b20670677f75423012e573a1550/src/AElf.Kernel.SmartContract.ExecutionPluginForResourceFee/ResourceTokenChargedLogEventProcessor.cs#L53-L58

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a `TODO` comment in d0f009f1d1ca9b20670677f75423012e573a1550 when #2611 was merged. cc @AElfProject.

### PR DESCRIPTION
**Pull request by EanCuznaivy**
Now every LogEventProcessor can handle interested log events of one block at once.

Fix #2632 Fix #2633

### CODE REVIEW

File: src/AElf.Kernel.Core/SmartContract/Application/LogEventListeningService.cs [hunk 1]
```diff
@@ -20,44 +20,36 @@ public class LogEventListeningService<T> : ILogEventListeningService<T>
...
                     {
+                        // No interested event in the block
                         continue;
                     }
 
-                    result.BlockHash = block.GetHash();
-                    var resultBloom = new Bloom(result.Bloom.ToByteArray());
-
-                    foreach (var processor in _logEventProcessors)
+                    foreach (var result in txResults.AsParallel())
```
Comment (loning) [inline src/AElf.Kernel.Core/SmartContract/Application/LogEventListeningService.cs]: remove .AsParallel()

File: src/AElf.Kernel.Core/SmartContract/Application/LogEventListeningService.cs [hunk 2]
```diff
@@ -70,15 +62,26 @@ public async Task ProcessAsync(IEnumerable<Block> blocks)
...
-                                continue;
-                            await processor.ProcessAsync(block, result, log);
+                            if (log.Address != interestedEvent.Address || log.Name != interestedEvent.Name) continue;
+                            if (logEventsMap.ContainsKey(result))
+                            {
+                                logEventsMap[result].Add(log);
+                            }
+                            else
+                            {
+                                logEventsMap[result] = new ConcurrentBag<LogEvent>
```
Comment (loning) [inline src/AElf.Kernel.Core/SmartContract/Application/LogEventListeningService.cs]: new ConcurrentBag<LogEvent>

File: src/AElf.Kernel.Consensus.AEDPoS/Application/SecretSharingInformationLogEventProcessor.cs [hunk 3]
```diff
@@ -34,9 +35,16 @@ public LogEvent InterestedEvent
...
 
-        public Task ProcessAsync(Block block, TransactionResult result, LogEvent logEvent)
+        public Task ProcessAsync(Block block, Dictionary<TransactionResult, List<LogEvent>> logEventsMap)
         {
-            _secretSharingService.AddSharingInformationAsync(logEvent);
+            foreach (var logEvents in logEventsMap.Values)
+            {
+                foreach (var logEvent in logEvents)
+                {
+                    _secretSharingService.AddSharingInformationAsync(logEvent);
```
Comment (loning) [inline src/AElf.Kernel.Consensus.AEDPoS/Application/SecretSharingInformationLogEventProcessor.cs]: _secretSharingService.AddSharingInformationAsync should not accept logEvent as parameters

File: src/AElf.Kernel.Consensus.AEDPoS/AssemblyInfo.cs [hunk 4]
```diff
@@ -3,3 +3,5 @@
 [assembly: InternalsVisibleTo("AElf.Kernel.Tests")]
 [assembly: InternalsVisibleTo("AElf.Contracts.Consensus.AEDPoS.Tests")]
 [assembly: InternalsVisibleTo("AElf.Contracts.Economic.TestBase")]
+[assembly: InternalsVisibleTo("AElf.Contracts.TestKit.AEDPoSExtension")]
```
Comment (loning) [inline src/AElf.Kernel.Consensus.AEDPoS/AssemblyInfo.cs]: do not use it except test projects

Comment (loning) [inline src/AElf.Kernel.Consensus.AEDPoS/AssemblyInfo.cs]: do not use it except test projects

Comment (EanCuznaivy) [reply src/AElf.Kernel.Consensus.AEDPoS/AssemblyInfo.cs]: > do not use it except test projects

This is a project act as test base, like testkit.

**Comment by loning** [comment]
ok

EanCuznaivy <notifications@github.com>于2020年3月31日 周二15:45写道：

**Pull request merged**